core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/create/TransportCreateSnapshotAction.java

```diff
@@ -87,16 +85,16 @@ public void onResponse() {
                 if (request.waitForCompletion()) {
                     snapshotsService.addListener(new SnapshotsService.SnapshotCompletionListener() {
                         @Override
-                        public void onSnapshotCompletion(SnapshotId snapshotId, SnapshotInfo snapshotInfo) {
-                            if (snapshot.equals(snapshotId.getSnapshot())) {
+                        public void onSnapshotCompletion(Snapshot snapshot, SnapshotInfo snapshotInfo) {
+                            if (snapshot.isSame(request.repository(), request.snapshot())) {
```

Comment:
Its not clear to me what the above API is intended to achieve?  As I see it, the listeners have this check in them because on snapshot/restore completion, every registered listener is looped through and `onCompletion` is called, so the check is to make sure the listener only truly runs on the right snapshot.  I think the listeners could be redone to be in a map instead of a list, where the map has entries e.g. `Snapshot -> List<SnapshotCompletionListener>`.  But that change seems beyond the scope of this PR.